File: packages/agent/src/infrastructure/events/types/worktree.ts
/**
 * @fileoverview Worktree Events
 *
 * Events for git worktree operations.
 */

import type { SessionId } from './branded.js';
import type { BaseEvent } from './base.js';

// =============================================================================
// Worktree Events
// =============================================================================

/**
 * Worktree acquired event - session has a working directory
 */
export interface WorktreeAcquiredEvent extends BaseEvent {
  type: 'worktree.acquired';
  payload: {
    /** Filesystem path to working directory */
    path: string;
    /** Git branch name */
    branch: string;
    /** Starting commit hash */
    baseCommit: string;
    /** Whether this is isolated (worktree) or shared (main directory) */
    isolated: boolean;
    /** If forked, the parent session's info */
    forkedFrom?: {
      sessionId: SessionId;
      commit: string;
    };
  };
}

/**
 * Worktree commit event - changes committed in session's worktree
 */
export interface WorktreeCommitEvent extends BaseEvent {
  type: 'worktree.commit';
  payload: {
    /** Git commit hash */
    commitHash: string;
    /** Commit message */
    message: string;
    /** Files changed in this commit */
    filesChanged: string[];
    /** Number of insertions */
    insertions?: number;
    /** Number of deletions */
    deletions?: number;
  };
}

/**
 * Worktree released event - session released its working directory
 */
export interface WorktreeReleasedEvent extends BaseEvent {
  type: 'worktree.released';
  payload: {
    /** Final commit hash (if changes were committed) */
    finalCommit?: string;
    /** Whether worktree was deleted */
    deleted: boolean;
    /** Whether branch was preserved */
    branchPreserved: boolean;
  };
}

/**
 * Worktree merged event - session's branch was merged
 */
export interface WorktreeMergedEvent extends BaseEvent {
  type: 'worktree.merged';
  payload: {
    /** Branch that was merged */
    sourceBranch: string;
    /** Target branch */
    targetBranch: string;
    /** Merge commit hash */
    mergeCommit: string;
    /** Merge strategy used */
    strategy: 'merge' | 'rebase' | 'squash';
  };
}

File: packages/agent-rs/crates/tron-events/src/types/payloads/worktree.rs
//! Worktree event payloads.

use serde::{Deserialize, Serialize};

/// Payload for `worktree.acquired` events.
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct WorktreeAcquiredPayload {
    /// Worktree path.
    pub path: String,
    /// Branch name.
    pub branch: String,
    /// Base commit hash.
    pub base_commit: String,
    /// Whether the worktree is isolated.
    pub isolated: bool,
    /// Fork source if applicable.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub forked_from: Option<WorktreeForkSource>,
}

/// Worktree fork source.
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct WorktreeForkSource {
    /// Source session ID.
    pub session_id: String,
    /// Source commit hash.
    pub commit: String,
}

/// Payload for `worktree.commit` events.
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct WorktreeCommitPayload {
    /// Commit hash.
    pub commit_hash: String,
    /// Commit message.
    pub message: String,
    /// Files changed.
    pub files_changed: Vec<String>,
    /// Lines inserted.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub insertions: Option<i64>,
    /// Lines deleted.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub deletions: Option<i64>,
}

/// Payload for `worktree.released` events.
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct WorktreeReleasedPayload {
    /// Final commit hash.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub final_commit: Option<String>,
    /// Whether the worktree was deleted.
    pub deleted: bool,
    /// Whether the branch was preserved.
    pub branch_preserved: bool,
}

/// Payload for `worktree.merged` events.
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct WorktreeMergedPayload {
    /// Source branch name.
    pub source_branch: String,
    /// Target branch name.
    pub target_branch: String,
    /// Merge commit hash.
    pub merge_commit: String,
    /// Merge strategy.
    pub strategy: String,
}
